File: app/src/main/java/com/hnu/heshequ/home/ui/HotActivitiesFragment.java
package com.hnu.heshequ.home.ui;

import android.os.Handler;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import com.google.gson.Gson;
import com.hnu.heshequ.R;
import com.hnu.heshequ.adapter.recycleview.HotActiveAdapter;
import com.hnu.heshequ.base.NetWorkFragment;
import com.hnu.heshequ.bean.ConsTants;
import com.hnu.heshequ.bean.HotAvtivityBean;
import com.hnu.heshequ.constans.Constants;
import com.hnu.heshequ.utils.Utils;
import com.jcodecraeer.xrecyclerview.XRecyclerView;

import org.json.JSONObject;

import java.util.ArrayList;

public class HotActivitiesFragment extends NetWorkFragment implements XRecyclerView.LoadingListener {

    private static final String TAG = "[HotActivityFragment]";
    private HotActiveAdapter adapter;
    private TextView tvTips;
    XRecyclerView mRecyclerView;
    private final int GetData = 1000;
    private final int LoaMore = 1001;

    private int pn;
    private int totalPage;
    private HotAvtivityBean hotAvtivityBean;
    private final Gson gson = new Gson();
    private ArrayList<HotAvtivityBean.HotBean> data;

    @Override
    protected View createView(LayoutInflater inflater) {
        View view = inflater.inflate(R.layout.fragment_tim, null);
        mRecyclerView = view.findViewById(R.id.rv);
        ConsTants.initXRecycleView(getActivity(), true, false, mRecyclerView);
        mRecyclerView.setLoadingListener(this);
//        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        tvTips = view.findViewById(R.id.tvTips);
        mRecyclerView.setNestedScrollingEnabled(false);

        getData(true);

        adapter = new HotActiveAdapter(getActivity());
        mRecyclerView.setAdapter(adapter);
        return view;
    }

    @Override
    public void onRefresh() {
//        new Handler().postDelayed(() -> mRecyclerView.refreshComplete(), 1000);
    }

    @Override
    public void onLoadMore() {
        new Handler().postDelayed(() -> getData(false), 1000);
    }

    public void getData(boolean isRefresh) {
        if (isRefresh) {
            pn = 1;
            setBodyParams(new String[]{"pn", "ps"}, new String[]{"" + pn, Constants.default_PS + ""});
            sendPostConnection(Constants.base_url + "/api/club/activity/hotlist.do", GetData, Constants.token);
        } else if (pn < totalPage) {
            pn++;
            setBodyParams(new String[]{"pn", "ps"}, new String[]{"" + pn, Constants.default_PS + ""});
            sendPostConnection(Constants.base_url + "/api/club/activity/hotlist.do", LoaMore, Constants.token);
        } else {
            mRecyclerView.loadMoreComplete();
        }
    }

    @Override
    protected void onSuccess(JSONObject result, int where, boolean fromCache) {
        Log.d(TAG, "onSuccess: where: " + where + ", result: " + result.toString());
        if (result.optInt("code") != 0) {
            Utils.toastShort(mContext, result.optString("msg"));
            return;
        }
        switch (where) {
            case GetData:
                hotAvtivityBean = gson.fromJson(result.optString("data"), HotAvtivityBean.class);
                totalPage = hotAvtivityBean.getTotalPage();
                if (hotAvtivityBean.getData() == null) {
                    return;
                }
                if (!hotAvtivityBean.getData().isEmpty()) {
                    tvTips.setVisibility(View.GONE);
                    data = hotAvtivityBean.getData();
                    adapter.setData(data);
                } else {
                    tvTips.setVisibility(View.VISIBLE);
                    data = new ArrayList<>();
                    adapter.setData(data);
                }
                break;
            case LoaMore:
                mRecyclerView.loadMoreComplete();
                hotAvtivityBean = gson.fromJson(result.optString("data"), HotAvtivityBean.class);
                totalPage = hotAvtivityBean.getTotalPage();
                if (hotAvtivityBean.getData() == null) {
                    return;
                }
                if (!hotAvtivityBean.getData().isEmpty()) {
                    data = hotAvtivityBean.getData();
                    adapter.setData2(data);
                } else {
                    data = new ArrayList<>();
                    adapter.setData2(data);
                }
                break;
        }
    }

    @Override
    protected void onFailure(String result, int where) {
        Utils.toastShort(mContext, "网络异常");
    }
}

File: app/src/main/java/com/hnu/heshequ/home/ui/HotActivitiesFragment.kt
package com.hnu.heshequ.home.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.hnu.heshequ.R
import com.hnu.heshequ.adapter.recycleview.HotActiveAdapter
import com.hnu.heshequ.bean.ConsTants
import com.hnu.heshequ.home.model.HotActivitiesViewModel
import com.jcodecraeer.xrecyclerview.XRecyclerView
import com.jcodecraeer.xrecyclerview.XRecyclerView.LoadingListener
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class HotActivitiesFragment : Fragment(), IListFragment {

    private lateinit var recyclerView: XRecyclerView
    private var adapter: HotActiveAdapter? = null
    private lateinit var tvTips: TextView

    private val viewModel: HotActivitiesViewModel by viewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val view = inflater.inflate(R.layout.fragment_tim, container, false)
        recyclerView = view.findViewById(R.id.rv)
        recyclerView.setLoadingListener(object : LoadingListener {
            override fun onRefresh() {
            }

            override fun onLoadMore() {
                viewLifecycleOwner.lifecycleScope.launch {
                    delay(1000)
                    viewModel.fetchData()
                }
            }
        })
        recyclerView.isNestedScrollingEnabled = false
        ConsTants.initXRecycleView(activity, true, false, recyclerView)
        adapter = HotActiveAdapter(activity)
        recyclerView.adapter = adapter

        tvTips = view.findViewById(R.id.tvTips)

        initCollect()
        return view
    }

    private fun initCollect() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.hotActivitiesBeanStateFlow
                .collect {
                    tvTips.visibility = if (it.isNotEmpty()) View.GONE else View.VISIBLE
                    recyclerView.refreshComplete()
                    adapter?.setData(it)
                }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.loadFinish.collect {
                recyclerView.loadMoreComplete()
            }
        }
    }

    override fun isFirstItemVisible(): Boolean {
        val layoutManager = recyclerView.layoutManager as LinearLayoutManager?
        return if (layoutManager != null) {
            layoutManager.findFirstCompletelyVisibleItemPosition() == 0
        } else {
            false
        }
    }

    companion object {
        private const val TAG = "[HotActivitiesFragment]"
    }
}